hasher-matcher-actioner/hmalib/lambdas/unified_indexer.py

```diff
@@ -0,0 +1,172 @@
+    HashRowT,
+    ThreatExchangeS3PDQAdapter,
+    ThreatExchangeS3VideoMD5Adapter,
+    ThreatUpdateS3Store,
+    S3ThreatDataConfig,
+)
+from hmalib.common.logging import get_logger
+from hmalib import metrics
+
+logger = get_logger(__name__)
+
+THREAT_EXCHANGE_DATA_BUCKET_NAME = os.environ["THREAT_EXCHANGE_DATA_BUCKET_NAME"]
+THREAT_EXCHANGE_DATA_FOLDER = os.environ["THREAT_EXCHANGE_DATA_FOLDER"]
+
+INDEXES_BUCKET_NAME = os.environ["INDEXES_BUCKET_NAME"]
+
+
+def unwrap_if_sns(data):
+    """
+    If SNS message, extracts the message. Else, returns orignal message itself.
+    """
+
+    if "EventSource" in data and data["EventSource"] == "aws:sns":
+        message = data["Sns"]["Message"]
+        return json.loads(message)
+    return data
+
+
+def is_s3_testevent(data):
+    return "Event" in data and data["Event"] == "s3:TestEvent"
+
+
+def get_updated_files_by_signal_type(event) -> t.Dict[t.Type[SignalType], t.List[str]]:
+    """
+    Given SQS events, creates map of signal_type -> list(updated_files). Returns
+    empty map if none of the updated files have signals. Empty maps are Falsy.
+
+    There are two reasons why we may receive batches.
+    - SQS attribute batch_window_in_seconds can be kept large in this case to
+      allow batches to form
+    - files are likely going to be updated en-masse when /threat_updates are
+      received.
+    """
+    result = defaultdict(list)
+    for record in event["Records"]:
+        inner_record = unwrap_if_sns(record)
+        if is_s3_testevent(inner_record):
+            continue
+        for s3_record in inner_record["Records"]:
+            bucket_name = s3_record["s3"]["bucket"]["name"]
+            file_path = unquote_plus(s3_record["s3"]["object"]["key"])
+            if (
+                bucket_name == THREAT_EXCHANGE_DATA_BUCKET_NAME
+                and file_path.startswith(THREAT_EXCHANGE_DATA_FOLDER)
+            ):
+                optional_signal_type = (
+                    ThreatUpdateS3Store.get_signal_type_from_object_key(file_path)
+                )
+                if optional_signal_type:
+                    result[optional_signal_type].append(file_path)
+
+    return result
+
+
+def merge_threat_exchange_files(
+    accumulator: t.Dict[str, HashRowT], hash_row: HashRowT
+) -> t.Dict[str, HashRowT]:
+    hash, meta_data = hash_row
+    if hash not in accumulator.keys():
+        # Add hash as new row
+        accumulator[hash] = hash_row
+    else:
+        # Add new privacy group to existing row
+        accumulator[hash][1]["privacy_groups"].update(meta_data["privacy_groups"])
+        # Add new 'tags' for privacy group to existing row
+        accumulator[hash][1]["tags"].update(meta_data["tags"])
+    return accumulator
+
+
+# Maps from signal type to the subclass of ThreatExchangeS3Adapter.
+# ThreatExchangeS3Adapter is used to fetch all the data corresponding to a
+# signal_type. At some point, we must allow _updates_ to indexes rather than
```

Comment:
irrc for FAISS the only 'efficient' update is a delete otherwise rebuild is just as fast in most cases.